### PR TITLE
Create weekly event recurring instance

### PR DESCRIPTION
**Pull request by Community-Programmer**
**What kind of change does this PR introduce?**
Fixed the bug on creating recurring event only single event is created

**Issue Number:**

Fixes https://github.com/PalisadoesFoundation/talawa-api/issues/1583

**Did you add tests for your changes?**
yes

**Snapshots/Videos:**

https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/2d723ff1-ae9d-4bfa-a6a7-a7a0597d4c87

**If relevant, did you update the documentation?**

No

**Summary**

- [ ]  Create a weekly recurring instance

**Implementation Details:**

**Validation:**

No other functionalities is affected by changes made in the createEvent  mutation
All test files passed
image

**Does this PR introduce a breaking change?**

Yes

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Community-Programmer** [comment]
@palisadoes @sumitra19jha @ttaylor92 

### Attention:

I have made the required changes, but the test case is failing because, as suggested by @sumitra19jha, 'to start a transaction and make sure everything is a success or everything fails.' 
Transactions are available for replica sets, not for standalone mongod instances.

image


So, the test case will fail if we are using a standalone MongoDB for development and testing.

If we want to work with transactions in MongoDB, we'll need to set up a replica set even for our development environment.

[Convert a Standalone mongod to a Replica Set](https://www.mongodb.com/docs/manual/tutorial/convert-standalone-to-replica-set/#:~:text=A%20standalone%20mongod%20instance%20is,cluster%2C%20provides%20redundancy%20and%20availability.)

All test cases are passing when using replica sets and MongoDB Atlas

Is there any other approach we should follow or continue with this?

**Comment by palisadoes** [comment]
@Community-Programmer @sumitra19jha 

1. Though it is a good idea, implementing DB replication is out of scope of this PR
2. We can create another issue to get that implemented after this PR is merged
3. Please assume no replication for all tests related to this PR

I'd like to get this merged so that we can start creating issues in Admin and Mobile to implement weekly recurring meetings in their respective UIs

**Comment by Community-Programmer** [comment]
> @Community-Programmer @sumitra19jha
> 
> 1. Though it is a good idea, implementing DB replication is out of scope of this PR
> 2. We can create another issue to get that implemented after this PR is merged
> 3. Please assume no replication for all tests related to this PR
> 
> I'd like to get this merged so that we can start creating issues in Admin and Mobile to implement weekly recurring meetings in their respective UIs

can we comment out or skip the tests related to database replication for now? 

I'll make sure to include clear comments in the code, explicitly stating that these tests are skipped temporarily until database replication is implemented.

This approach ensures that these tests are documented within the PR, and we can easily reintroduce them once the database replication feature is implemented.

**Comment by palisadoes** [comment]
> > @Community-Programmer @sumitra19jha
> > 
> > 1. Though it is a good idea, implementing DB replication is out of scope of this PR
> > 2. We can create another issue to get that implemented after this PR is merged
> > 3. Please assume no replication for all tests related to this PR
> > 
> > I'd like to get this merged so that we can start creating issues in Admin and Mobile to implement weekly recurring meetings in their respective UIs
> 
> can we comment out or skip the tests related to database replication for now?
> 
> I'll make sure to include clear comments in the code, explicitly stating that these tests are skipped temporarily until database replication is implemented.
> 
> This approach ensures that these tests are documented within the PR, and we can easily reintroduce them once the database replication feature is implemented.

1. Yes, that is acceptable
1. @sumitra19jha @xoldyckk Please review

File: src/resolvers/Mutation/createEvent.ts
Comment (sumitra19jha) [inline src/resolvers/Mutation/createEvent.ts:123]: So your plan is to start session every time which is bad implementation. Please start one session and keep sing it for transaction throught the server

**Review by sumitra19jha — CHANGES_REQUESTED** [review]

File: src/resolvers/Mutation/createEvent.ts
Comment (Community-Programmer) [reply src/resolvers/Mutation/createEvent.ts]: Thank you for your review and feedback. I have taken your suggestion into consideration and implemented a single session for the entire application, now a session is initiated when the database connection is established

**Comment by Community-Programmer** [comment]
@sumitra19jha @xoldyckk @ttaylor92

Please review

**Comment by sumitra19jha** [comment]
Please write the code clean and correctly. Have you run the code to see if transactions are working? What do you mean by commenting out for replica? If you need replica, try to resolve it locally and run it there. If you don't, then don't put this unnecessary comments.


Also please note when you ask for review, you have provided a clean code, containing logic, docs and test cases. Every mentor is busy, so when they put their time, they are looking for reviewing your final solution which might need few modifications (in another review) and then done.

**Comment by Community-Programmer** [comment]
I have completely tested the code before submitting the PR. I tested each functionality locally including working of transactions and converted the standalone to replica locally for testing this, also tested it using MongoDB Atlas. I have commented out the code because if this gets merged it will cause trouble for others as test case will fail for them .So I commented out these test cases so that it can be reintroduced in the future once DB replication is implemented.

Also i am documenting all the changes I am doing so that it help others when integrating this feature with Talawa admin and mobile

In terms of cleaner code, I tried my best to reduce complexity and make it as understandable as possible. If you want, I can further clean it up by moving helper functions into a separate file. Further the complexity will increase when dealing with monthly recurring events.

I apologise if asking for a review annoyed you. I understand you are busy reviewing other PRs. I'm sorry for my behavior

Thank you

**Review by palisadoes — CHANGES_REQUESTED** [review]
In hindsight, you have commented out test cases and stated that the code is thoroughly tested. The code coverage has declined significantly which is not desirable. This implies the testing done is manual, which is unverifiable. The code has therefore become more unstable.

If your testing logic requires replication to be enabled that means all other tests in the code base are at risk of being removed. This is not acceptable.

Restore the tests and figure out how to verify the functionality your new code thoroughly in an automated way.

**Comment by palisadoes** [comment]
I'd interpret `to start a transaction and make sure everything is a success or everything fails` as meaning that there is transaction rollback capability with error handling that an ODM/ORM like Mongoose should support.

**Comment by sumitra19jha** [comment]
> I'd interpret `to start a transaction and make sure everything is a success or everything fails` as meaning that there is transaction rollback capability with error handling that an ODM/ORM like Mongoose should support.

Exactly correct. Its very simple
1. Start a transaction 
2. Perform CRUD
3. Commit the transaction 
4. If it fails rollback

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:327]: Why hasn't this code been restored given the past comments by me and @sumitra19jha ?

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comment

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts]: sorry , I removed the commit due to some format issues.

**Review by palisadoes — DISMISSED** [review]

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by Community-Programmer** [comment]
> Please fix the failing tests

Fixed all the failing test files

**Comment by palisadoes** [comment]
Is this ready for review now?

**Comment by Community-Programmer** [comment]
> Is this ready for review now?

Yaa I hope so, I made all the suggest changes given by @sumitra19jha 

- I have Optimized the code and written it in much cleaner way
- Properly tested and written test files for the same

I will complete the documentation once it get reviewed.

**Comment by palisadoes** [comment]
Do the documentation as part of this PR

**Comment by palisadoes** [comment]
@sumitra19jha @xoldyckk Please review

**Comment by Community-Programmer** [comment]
> Do the documentation as part of this PR

Ok sir, sure

**Comment by Community-Programmer** [comment]
I have fixed the linting issues and resolved merge conflicts

**Comment by palisadoes** [comment]
@sumitra19jha 
@xoldyckk 
@DMills27 
@JamaicanFriedChicken 

Please review

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:188]: Will this test work after `2023-01-29T00:00:00Z`?

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:196]: Will this test work after `2023-01-29T00:00:00Z`?

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:199]: Would enums be better here?

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:277]: Will this test work after this date?

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (palisadoes) [inline tests/resolvers/Mutation/createEvent.spec.ts:367]: Will this test work after `2023-01-29T00:00:00Z`?

File: src/helpers/eventInstance.ts
Comment (palisadoes) [inline src/helpers/eventInstance.ts:45]: Where is the logic for no end date? It was supposed to January 1, 2100

**Comment by palisadoes** [comment]
Please take a look:

1. @xoldyckk
1. @sumitra19jha
1. @DMills27
1. @JamaicanFriedChicken

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts:188]: Yes, it will work for any date. This is the  sample date that  was used to create multiple events for verification, but the test is not restricted to that specific date and will function for any date provided.
I have taken this so that n number of events are generated and same is verified

image

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts:196]: Same goes to this

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts:199]: No it should be string only we used enumerations to define  named constant values. In our case, the recurrance property is expected to have one of the following values: 
"ONCE", "DAILY", "WEEKLY", "MONTHLY","YEARLY".
so we have to write test cases for each

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts:277]: same here

File: tests/resolvers/Mutation/createEvent.spec.ts
Comment (Community-Programmer) [reply tests/resolvers/Mutation/createEvent.spec.ts:367]: same here

File: src/helpers/eventInstance.ts
Comment (Community-Programmer) [reply src/helpers/eventInstance.ts:45]: I have not added it because I believe it is not required. By default in our model, if `startDate` is given, `endDate` should also be required. I consider this a good practice.

Since we are following Google Calendar, where the end date is also required, the same principle applies to Talawa-admin/mobile. We do not provide users with an option to not give the endDate.

Consider creating `Daily` recurring events with no end date ; it will generate approximately 27,000 events.

**Comment by Community-Programmer** [comment]
@palisadoes 

resolved conversation please do see the comments once

**Comment by palisadoes** [comment]
How do you plan to handle custom intervals like:

1. Weekly on Tuesday and Thursday
1. Third Monday of the month?
1. Last Friday of the month
1. Monthly on the first day

Where would that logic our in future?

File: src/helpers/eventInstance.ts
Comment (palisadoes) [inline src/helpers/eventInstance.ts:2]: How can this file be refactored so that functions to handle different intervals can be worked on simultaneously by multiple contributors minimizing merge conflicts?

We'll need to do this to speed up development time. Make a proposal because this approach will need to be done and is stated in the original issue

**Review by palisadoes — CHANGES_REQUESTED** [review]
How can the `eventInstance.ts` file be refactored so that functions to handle different intervals can be worked on simultaneously by multiple contributors minimizing merge conflicts?

We'll need to do this to speed up development time. Make a proposal because this approach will need to be done and is stated in the original issue

**Comment by Community-Programmer** [comment]
@palisadoes 

**For regular intervals** - I have already implemented the functionality for intervals as described below. The backend is set up to manage the scheduling of events based on the specified intervals. 
For instance, if today is 12/01/2024, and it's Wednesday, and a weekly event is created, events will be generated every week on Wednesdays.

See below event is created `weekly on monday`

https://github.com/PalisadoesFoundation/talawa-api/assets/76515568/b6a822e1-2dd1-41d4-9992-9f34463add72


Similarly, I've implemented functionality for events occurring on the first Wednesday, last Wednesday, third Wednesday, etc. The backend takes care of these calculations. Now, we need to work on displaying the corresponding custom messages on the frontend.

**For custom intervals** - I have thought of using two npm modules one `momentjs` and one` date-fns`
https://date-fns.org
https://momentjs.com
this module will help us to get the required date we want to create custom events.

The other functionality is almost complete, and I am waiting for this pull request to be thoroughly reviewed and merged. This will help me become familiar with it and serve as the foundation for all my future work, making it easier for other pull requests to be reviewed

File: src/helpers/eventInstance.ts
Comment (Community-Programmer) [reply src/helpers/eventInstance.ts:2]: @palisadoes 

Ok we can refactor this like

 **eventInstance/
  ├── index.ts
  ├── daily.ts
  ├── weekly.ts
  └── monthly.ts**

I have completed all the features, and now I only need the pull requests to be reviewed quickly. If so, I believe we can finish all the features in a very short amount of time

**Comment by Community-Programmer** [comment]
@palisadoes 

I am refactoring the file and also switching to `moment js` and `date-fns` module so that overall work can be speed up

refactoring like this

**eventInstance/
├── index.ts
├── daily.ts
├── weekly.ts
└── monthly.ts**

**Comment by palisadoes** [comment]
Please take a look:

1. @xoldyckk
1. @sumitra19jha
1. @DMills27
1. @JamaicanFriedChicken

This is ready for review

**Comment by palisadoes** [comment]
> @palisadoes 
> 
> I am refactoring the file and also switching to `moment js` and `date-fns` module so that overall work can be speed up
> 
> refactoring like this
> 
> **eventInstance/
> ├── index.ts
> ├── daily.ts
> ├── weekly.ts
> └── monthly.ts**

Yes, that's the better approach

**Review by JamaicanFriedChicken — DISMISSED** [review]
It looks good!

**Comment by Community-Programmer** [comment]
Sorry Closed wrongly

**Comment by palisadoes** [comment]
1. The code seems create N number of Event entries for N weeks.
2. How will someone be able to edit the series? It's not clear. Will that be a separate PR?

**Comment by Community-Programmer** [comment]
@palisadoes 

Regarding the editing of events in a series, the plan is to introduce that functionality towards the end of the development process. Initially, we'll focus on adding features for each event creation within the series. Once those features are in place, we'll then extend the functionality to allow for editing the entire series

**flow for editing series of event will be like this:**

we'll implement a feature where we fetch events based on a specified start date and end date.Also When creating a series of events, I'll include an edit in the event model by adding a new field called `recurringId `. All recurring events in the same series will have a unique `recurringId`. This approach will make it easier to identify and edit the entire series of events by referencing the common `recurringId`.

**Comment by palisadoes** [comment]
OK. Merging

**Comment by palisadoes** [comment]
Adding for cross referencing
- https://github.com/PalisadoesFoundation/talawa-api/pull/1712

**Comment by DecodeAndCode** [comment]
@Community-Programmer I am not able use this weekly recurring property even after adding a sample data in db for a weekly recurring event but in admin its showing its occurance as one. can you please me out here??

**Pull request merged**
